### DISCUSSION TITLE
SceneGraphLayer getOrientation/getTransformMatrix attribute buffer

### DISCUSSION THREAD

**Discussion by samelie** [visgl/deck.gl#5801] [Q&A]
Hello DeckGL'ers

Trying to supply the `getOrientation` accessor with a buffer but am failing.

`getPosition` is working fine with the buffer (am inspired by the [interleaved buffer example](https://github1s.com/visgl/deck.gl/blob/HEAD/examples/experimental/interleaved-buffer/app.js)) 

Looking at the source for SceneGraphLayer, it would seem `getPosition` and `getColor` are treated differently in the [`initializeState`](https://github1s.com/visgl/deck.gl/blob/HEAD/modules/mesh-layers/src/scenegraph-layer/scenegraph-layer.js#83) , where the model transformation go through another process.

I'd like to supply everything as array buffers if possible. The following does _not work_ 

```

const positions = { buffer: pBuffer, type: GL.FLOAT, size: 3, offset: 0, stride: 12 };
    const orientation = { buffer: oBuffer, type: GL.FLOAT, size: 3, offset: 0, stride: 12 };
    const transformation = { buffer: tBuffer, type: GL.FLOAT, size: 3, offset: 0, stride: 64 };

    const layer = new ScenegraphLayer({
        id: "assets",
        numInstances: numInstances,
        data: {
            attributes: {
                getPosition: positions,
                getTransformMatrix: transformation
                getOrientation: orientation
            }
        }
      }
      
```

If that is not possible can I mix _attributes_ and typical data accessors ( `{data:[{orientation:[0,0,0]}]}`} ) ?

Any guidance would be appreciated

**Comment by ibgreen** [comment]
My knowledge could be outdated, but perhaps these notes can help with your investigations:

- I think the shader only accepts a transformation matrix. So you either provide a per row transformation matrix or a per row orientation (in the latter case a per row transformation matrix is then built under the hood from orientation etc).
- Therefore, I would not expect binary buffers to work for orientation as they would first need to be composed into a single binary matrix attribute, and I am not sure this is currently being done for binary attributes
- You might be able to precompose and supply a binary matrix attribute, though there could be some additional problems to work through as a matrix attribute has to be spread over multiple GPU attributes (since each attribute can only hold 4 float32s). You can check the layer code to see how this is done, you may be able to replicate it.
- Also I seem to recall the matrix attribute is a 4x3 matrix (rather than a 4x4 matrix) in order to save space and reduce the number of attributes required.

**Reply by samelie** [reply]
Thank you Ib for this write up. Looks like piecing together various layers and their shaders for a custom implementation is the path. By way of that, hopefully I'll learn more about the AttributeManager; seems like a core piece that I'm fuzzy on.

I'll post back here for posterity.
